### ISSUE TITLE
Using websockify with NAT for multiple VNC servers across different PCs

### ISSUE THREAD

**Issue by nikoladsp** [novnc/websockify#321]
Hello,
I have a question regarding usage of websockify with NAT: say that I need to have a publicly available HTTPS server which sits in front of the NAT clients. In this scenario, one NAT client is actually PC with running VNC server and there can be many clients like this. Lets say that adding NAT client(s) does not have to be 'dynamic' (on the fly) - it is desirable, but not mandatory. 

Is it possible (and if yes - how?) that users from the outside world, sees only HTTPS server (some login page) and depending on selected NAT client (VNC server) be connected to it. I manage to do this using a port from the pool of opened ports in the given range, so external user may connect to VNC behind NAT using websocket. 

My client however is concerned that this is a security breach (due multiple opened ports for websockify to use) and would prefer that we use only one websockify port to achieve this. Plus, that port must not be available from the public network. In short, he would prefer that 443 is used for both HTTPS server (NGINX) and also websockify to target multiple VNC servers behind NAT. Additionally, multiple users can access such system at the same time. I, on the other hand am not convinced that this is possible (at least part considering sharing 443 between HTTPS server and websockify). 

I tried to find a solution on google, but the closest thing I found is to use HTTPS server that targets multiple VNCs by tampering systemd processes - but even in this solution, different ports on HTTPS server were used for different VNC servers (each target VNC server has dedicated VNC port at HTTPS server PC); and number of possible VNC clients is measured in 100s, so extending number of daemons that runs  websockify is not practical (at least I think so), plus, most of the time users are idle and not using a connection.

Thank you all in advance.

Best regards

**Comment by kanaka** [comment]
You can definitely serve web and websocket connections on the same port. This is the reason why WebSocket connection begin with a HTTP handshake that is then upgraded to a full websocket connection. However, your web server does need to support WebSocket upgrades. Many popular web servers have had this support for a while including NGINX. Even websockify has an embedded web server that can turned on and it supports WebSocket upgrades although this is not recommended for production.

In terms of targeting multiple VNC servers, websockify supports a token based system for selecting between multiple targets: https://github.com/novnc/websockify/wiki/Token-based-target-selection 

In terms of security model, I would recommend that you only enable WebSocket upgrades for encrypted websocket (wss, port 443) to prevent eves dropping. You also need to consider how you will prevent unauthorized users from guessing or discovering a token and being able to connect via WebSockets to one of your VNC servers. This is security infrastructure that you will need to configure and/or build. Often this is done with a web UI that uses a well know authorization mechanism and then only enables WebSocket upgrades for that connection. Another option is for the web UI to create a short-lived websockify connection token that is removed once a connection is made. Hope that helps.

**Comment by nikoladsp** [comment]
Thank you for you comments. I saw Kimchi is using Tokens but I did not assume it has anything with 'multiplexing' the connection; so I will definitely look that topic. 

Also I have to search NGINX settings to check shall I do any additional steps regarding WebSocket upgrades. Do I need reverse proxy for this kind of scenario or token-base selection is enough?

Application is Flask python and uses JWT tokens so I don't think it will be a hassle to extend authorization.

Many thanks one more time for your effort!
